# Train model

In [1]:
import math
import numpy as np
import os
from pathlib import Path

from lapsim.encoder.partition import Partition
from lapsim.logging.logger import NewLogger

from lapsim.normalisation import TransformNormalisation

BATCH_SIZE = 256
EPOCHS = 200
FORESIGHT = 120
SAMPLING = 4
NORMAL_SPACING = 10
CHECKPOINT_EVERY = 5
NORMALISATION_BOUNDS_PATH = f"bounds-{NORMAL_SPACING}.json"

DATA_PATH = Path(fr"../dataset/dataset/encoded-{NORMAL_SPACING}/train")

# This code currently assumes you have one validation partition. but should be easy to add more
VAL_DATA_PATH = Path(fr"../dataset/dataset/encoded-{NORMAL_SPACING}/val")

# For test data
SPLICED_DATA_PATH = Path(fr"../dataset/dataset/spliced-{NORMAL_SPACING}/real-test")
TEST_DATA_PATH = Path(fr"../dataset/dataset/encoded-{NORMAL_SPACING}/real-test")

TENSOR_CACHE_DIR = fr"/run/media/belle/Development/Tensors-120f-{NORMAL_SPACING}m"

TRAINING_PARTITIONS = [x for x in os.listdir(DATA_PATH) if x[0] != '.']
VALIDATION_PARTITIONS = [x for x in os.listdir(VAL_DATA_PATH) if x[0] != '.']
TEST_PARTITIONS = [x for x in os.listdir(TEST_DATA_PATH) if x[0] != '.']


In [2]:
bounds = TransformNormalisation()

if os.path.exists(NORMALISATION_BOUNDS_PATH):
    bounds = TransformNormalisation.load(NORMALISATION_BOUNDS_PATH)
    print("Existing bounds loaded.")

else:
    print("No bounds file found, calculating new ones.")
    partitions = os.listdir(DATA_PATH)

    for p in TRAINING_PARTITIONS:
        partition = Partition.load(DATA_PATH / p)
        bounds.extend(partition)

    bounds.save(NORMALISATION_BOUNDS_PATH)
    print("Finished calculating bounds.")

bounds.transform.method = "flat-window"
bounds.transform.foresight = FORESIGHT
bounds.transform.sampling = SAMPLING


Existing bounds loaded.


In [3]:
from lapsim.models.lapsim_dense import LapSimModelDense
import torch

def get_device():
    if torch.cuda.is_available():
        return torch.device("cuda")
    elif torch.backends.mps.is_available():
        return torch.device("mps")

    return torch.device("cpu")


def tensor(x):
    return torch.tensor(x, dtype=torch.float32).to(device)


device = get_device()

# model = LapSimTransformerModel().to(device)
model = LapSimModelDense().to(device)
total_params = sum(p.numel() for p in model.parameters())
print("Params", total_params)

Params 547418


In [ ]:
from lapsim.evals.evaluation import Evaluation
from lapsim.evals import evals_suite


def batchify(count, n):
    indexes = np.array(list(range(count)))
    np.random.shuffle(indexes)

    batches = []
    for i in range((count // n) + 1):
        batch = indexes[i * n:(i + 1)*n]
        if len(batch) > 0:
            batches.append(batch)

    return batches


best_loss = math.inf
best_model_perf = None

new_logger = NewLogger("lapsim", run_name="test")

x, (y_pos, y_vel), vehicles = bounds.normalise_and_transform(
    Partition.load(DATA_PATH / TRAINING_PARTITIONS[0]),
    cache_tensor_prefix=f"{TENSOR_CACHE_DIR}/{0}",
    cores=8
)

val_x, (val_y_pos, val_y_vel), val_vehicles = bounds.normalise_and_transform(
    Partition.load(VAL_DATA_PATH / VALIDATION_PARTITIONS[0]),
    cache_tensor_prefix=f"{TENSOR_CACHE_DIR}/val",
    cores=8)

for epoch in range(EPOCHS):
    # Train model
    model.train()
    for i, partition in enumerate(TRAINING_PARTITIONS):
        # Start preloading the next batch of data while model trains
        loader = bounds.async_load_and_normalise_partition(
            DATA_PATH / TRAINING_PARTITIONS[(i + 1) % len(TRAINING_PARTITIONS)],
            cache_tensor_prefix=f"{TENSOR_CACHE_DIR}/{i + 1}",
            cores=8)

        batches = batchify(x.shape[0], BATCH_SIZE)
        for batch_idx, batch in enumerate(batches):
            model.optimiser.zero_grad()

            pred_pos, pred_vel = model(tensor(x[batch]), tensor(vehicles[batch]))

            model.optimiser.zero_grad()
            pos_loss = model.loss(pred_pos, tensor(y_pos[batch]))
            vel_loss = model.loss(pred_vel, tensor(y_vel[batch]))
            total_loss = (2 * pos_loss + vel_loss) / 3
            total_loss.backward()
            model.optimiser.step()

            new_logger.log_training_data(
                epoch,
                batch=batch_idx,
                n_batches=len(batches),
                partition=i,
                position_loss=pos_loss.item(),
                velocity_loss=vel_loss.item(),
            )

        # Once preloaded has finished we can set it to the next x/y datums
        loader.join()
        x, (y_pos, y_vel), vehicles = loader.normalisation

    # Validation
    model.eval()
    with torch.no_grad():
        batches = batchify(val_x.shape[0], BATCH_SIZE)
        for batch_idx, batch in enumerate(batches):
            val_pred_pos, val_pred_vel = model(tensor(val_x[batch]), tensor(val_vehicles[batch]))
            pos_loss = model.loss(val_pred_pos, tensor(val_y_pos[batch]))
            vel_loss = model.loss(val_pred_vel, tensor(val_y_vel[batch]))

            new_logger.log_validation_data(
                epoch,
                batch=batch_idx,
                n_batches=len(batches),
                position_loss=pos_loss.item(),
                velocity_loss=vel_loss.item(),
            )

    current_val_loss = np.mean(new_logger.val_pos_loss)
    best_validation_reached = best_model_perf is None or current_val_loss < best_model_perf

    new_logger.flush(epoch)

    if best_validation_reached:
        torch.save(model.state_dict(), "ls2.pt")
        torch.save(model.optimiser.state_dict(), "ls2-optim.pt")
        best_model_perf = current_val_loss

    if (epoch + 1) % CHECKPOINT_EVERY == 0:
        results = evals_suite.run_evals_suite(model, TEST_DATA_PATH, SPLICED_DATA_PATH)
        results = Evaluation.combine(results)

        print(
            "LapTime Error:",
            round(results.laptime.abs_error, 3),
            "Pos Error:",
            round(results.position.mean_absolute, 3),
            "Vel Error:",
            round(results.velocity.mean_absolute, 3)
        )

        new_logger.log_checkpoint(
            epoch,
            model,
            (val_x[0], val_vehicles[0]),
            results
        )


Epoch: 1, Position Loss: 0.013346, Velocity Loss: 0.000391, Position Loss: 0.009527, Velocity Loss: 0.000157        
New best val loss: 0.009684757969807833
4009 - Indycar (source from 2006) - CN Shanghai International Circuit - East Handling Circuit 2.json                       n                                 LapTime Error: 3.257 Pos Error: 1.148 Vel Error: 2.588
Epoch: 2, Position Loss: 0.011757, Velocity Loss: 0.000280, Position Loss: 0.010179, Velocity Loss: 0.000107        
Epoch: 3, Position Loss: 0.011551, Velocity Loss: 0.000266, Position Loss: 0.011073, Velocity Loss: 0.000178        
Epoch: 4, Position Loss: 0.011463, Velocity Loss: 0.000260, Position Loss: 0.010107, Velocity Loss: 0.000106        
Epoch: 5, Position Loss: 0.011432, Velocity Loss: 0.000260, Position Loss: 0.009766, Velocity Loss: 0.000108        
Epoch: 6, Position Loss: 0.011414, Velocity Loss: 0.000258, Position Loss: 0.010339, Velocity Loss: 0.000155        
Epoch: 7, Position Loss: 0.011409, Velocity Lo

```Epoch: 1, Position Loss: 0.014392, Velocity Loss: 0.000830, Position Loss: 0.011888, Velocity Loss: 0.000156
New best val loss: 0.012044078903272748
Epoch: 2, Position Loss: 0.010173, Velocity Loss: 0.000302, Position Loss: 0.010381, Velocity Loss: 0.000110
New best val loss: 0.01049167040764587
Epoch: 3, Position Loss: 0.009145, Velocity Loss: 0.000229, Position Loss: 0.010696, Velocity Loss: 0.000168
Epoch: 4, Position Loss: 0.008742, Velocity Loss: 0.000206, Position Loss: 0.009311, Velocity Loss: 0.000154
New best val loss: 0.009464911781833507
Epoch: 5, Position Loss: 0.008525, Velocity Loss: 0.000195, Position Loss: 0.009186, Velocity Loss: 0.000191
New best val loss: 0.009377222828334197
Epoch: 6, Position Loss: 0.008389, Velocity Loss: 0.000188, Position Loss: 0.008541, Velocity Loss: 0.000227
New best val loss: 0.008768312298343517
Epoch: 7, Position Loss: 0.008284, Velocity Loss: 0.000183, Position Loss: 0.007996, Velocity Loss: 0.000218
New best val loss: 0.008214128101826645
Epoch: 8, Position Loss: 0.008212, Velocity Loss: 0.000180, Position Loss: 0.009162, Velocity Loss: 0.000141
Epoch: 9, Position Loss: 0.008164, Velocity Loss: 0.000177, Position Loss: 0.009403, Velocity Loss: 0.000137
Epoch: 10, Position Loss: 0.008115, Velocity Loss: 0.000175, Position Loss: 0.010728, Velocity Loss: 0.000150
Epoch: 11, Position Loss: 0.008075, Velocity Loss: 0.000173, Position Loss: 0.009012, Velocity Loss: 0.000209
Epoch: 12, Position Loss: 0.008037, Velocity Loss: 0.000170, Position Loss: 0.010075, Velocity Loss: 0.000095
Epoch: 13, Position Loss: 0.007998, Velocity Loss: 0.000168, Position Loss: 0.009014, Velocity Loss: 0.000085
Epoch: 14, Position Loss: 0.007961, Velocity Loss: 0.000166, Position Loss: 0.008901, Velocity Loss: 0.000086
Epoch: 15, Position Loss: 0.007934, Velocity Loss: 0.000165, Position Loss: 0.008995, Velocity Loss: 0.000113
Epoch: 16, Position Loss: 0.007911, Velocity Loss: 0.000164, Position Loss: 0.008803, Velocity Loss: 0.000077
Epoch: 17, Position Loss: 0.007890, Velocity Loss: 0.000163, Position Loss: 0.007344, Velocity Loss: 0.000100
New best val loss: 0.007443627750035375
Epoch: 18, Position Loss: 0.007871, Velocity Loss: 0.000162, Position Loss: 0.009116, Velocity Loss: 0.000108
Epoch: 19, Position Loss: 0.007851, Velocity Loss: 0.000160, Position Loss: 0.008178, Velocity Loss: 0.000085
Epoch: 20, Position Loss: 0.007832, Velocity Loss: 0.000159, Position Loss: 0.007660, Velocity Loss: 0.000091
Epoch: 21, Position Loss: 0.007819, Velocity Loss: 0.000158, Position Loss: 0.007488, Velocity Loss: 0.000102
Epoch: 22, Position Loss: 0.007806, Velocity Loss: 0.000158, Position Loss: 0.008556, Velocity Loss: 0.000106
Epoch: 23, Position Loss: 0.007793, Velocity Loss: 0.000157, Position Loss: 0.009153, Velocity Loss: 0.000165
Epoch: 24, Position Loss: 0.007781, Velocity Loss: 0.000156, Position Loss: 0.008042, Velocity Loss: 0.000061
Epoch: 25, Position Loss: 0.007772, Velocity Loss: 0.000156, Position Loss: 0.008637, Velocity Loss: 0.000103
Epoch: 26, Position Loss: 0.007759, Velocity Loss: 0.000156, Position Loss: 0.007659, Velocity Loss: 0.000083
Epoch: 27, Position Loss: 0.007756, Velocity Loss: 0.000156, Position Loss: 0.007707, Velocity Loss: 0.000101
Epoch: 28, Position Loss: 0.007742, Velocity Loss: 0.000155, Position Loss: 0.007916, Velocity Loss: 0.000070
Epoch: 29, Position Loss: 0.007732, Velocity Loss: 0.000155, Position Loss: 0.007908, Velocity Loss: 0.000140
Epoch: 30, Position Loss: 0.007724, Velocity Loss: 0.000155, Position Loss: 0.008988, Velocity Loss: 0.000110
Epoch: 31, Position Loss: 0.007716, Velocity Loss: 0.000155, Position Loss: 0.008243, Velocity Loss: 0.000100
Epoch: 32, Position Loss: 0.007706, Velocity Loss: 0.000154, Position Loss: 0.007915, Velocity Loss: 0.000156
Epoch: 33, Position Loss: 0.007702, Velocity Loss: 0.000154, Position Loss: 0.009441, Velocity Loss: 0.000111
Epoch: 34, Position Loss: 0.007692, Velocity Loss: 0.000154, Position Loss: 0.007776, Velocity Loss: 0.000152
Epoch: 35, Position Loss: 0.007685, Velocity Loss: 0.000154, Position Loss: 0.007995, Velocity Loss: 0.000135
Epoch: 36, Position Loss: 0.007680, Velocity Loss: 0.000153, Position Loss: 0.008201, Velocity Loss: 0.000087
Epoch: 37, Position Loss: 0.007668, Velocity Loss: 0.000153, Position Loss: 0.008557, Velocity Loss: 0.000110
Epoch: 38, Position Loss: 0.007658, Velocity Loss: 0.000152, Position Loss: 0.007319, Velocity Loss: 0.000181
Epoch: 39, Position Loss: 0.007652, Velocity Loss: 0.000151, Position Loss: 0.007610, Velocity Loss: 0.000119
Epoch: 40, Position Loss: 0.007646, Velocity Loss: 0.000151, Position Loss: 0.008140, Velocity Loss: 0.000183
Epoch: 41, Position Loss: 0.007638, Velocity Loss: 0.000150, Position Loss: 0.006846, Velocity Loss: 0.000179
New best val loss: 0.00702526442182716
Epoch: 42, Position Loss: 0.007637, Velocity Loss: 0.000150, Position Loss: 0.007212, Velocity Loss: 0.000195
Epoch: 43, Position Loss: 0.007630, Velocity Loss: 0.000150, Position Loss: 0.006768, Velocity Loss: 0.000079
New best val loss: 0.006847250508144498
Epoch: 44, Position Loss: 0.007627, Velocity Loss: 0.000150, Position Loss: 0.007409, Velocity Loss: 0.000085
Epoch: 45, Position Loss: 0.007622, Velocity Loss: 0.000149, Position Loss: 0.007089, Velocity Loss: 0.000102
Epoch: 46, Position Loss: 0.007616, Velocity Loss: 0.000149, Position Loss: 0.007061, Velocity Loss: 0.000095
Epoch: 47, Position Loss: 0.007610, Velocity Loss: 0.000149, Position Loss: 0.006463, Velocity Loss: 0.000094
New best val loss: 0.006556672611623071
Epoch: 48, Position Loss: 0.007610, Velocity Loss: 0.000149, Position Loss: 0.006973, Velocity Loss: 0.000096
Epoch: 49, Position Loss: 0.007605, Velocity Loss: 0.000148, Position Loss: 0.007037, Velocity Loss: 0.000099
Epoch: 50, Position Loss: 0.007604, Velocity Loss: 0.000148, Position Loss: 0.006852, Velocity Loss: 0.000088```

Epoch: 1, Position Loss: 0.014392, Velocity Loss: 0.000830, Position Loss: 0.011888, Velocity Loss: 0.000156

New best val loss: 0.012044078903272748
Epoch: 2, Position Loss: 0.010173, Velocity Loss: 0.000302, Position Loss: 0.010381, Velocity Loss: 0.000110

New best val loss: 0.01049167040764587
Epoch: 3, Position Loss: 0.009145, Velocity Loss: 0.000229, Position Loss: 0.010696, Velocity Loss: 0.000168

Epoch: 4, Position Loss: 0.008742, Velocity Loss: 0.000206, Position Loss: 0.009311, Velocity Loss: 0.000154

New best val loss: 0.009464911781833507
Epoch: 5, Position Loss: 0.008525, Velocity Loss: 0.000195, Position Loss: 0.009186, Velocity Loss: 0.000191

New best val loss: 0.009377222828334197
Epoch: 6, Position Loss: 0.008389, Velocity Loss: 0.000188, Position Loss: 0.008541, Velocity Loss: 0.000227

New best val loss: 0.008768312298343517
Epoch: 7, Position Loss: 0.008284, Velocity Loss: 0.000183, Position Loss: 0.007996, Velocity Loss: 0.000218

New best val loss: 0.008214128101826645
Epoch: 8, Position Loss: 0.008212, Velocity Loss: 0.000180, Position Loss: 0.009162, Velocity Loss: 0.000141

Epoch: 9, Position Loss: 0.008164, Velocity Loss: 0.000177, Position Loss: 0.009403, Velocity Loss: 0.000137

Epoch: 10, Position Loss: 0.008115, Velocity Loss: 0.000175, Position Loss: 0.010728, Velocity Loss: 0.000150

Epoch: 11, Position Loss: 0.008075, Velocity Loss: 0.000173, Position Loss: 0.009012, Velocity Loss: 0.000209

Epoch: 12, Position Loss: 0.008037, Velocity Loss: 0.000170, Position Loss: 0.010075, Velocity Loss: 0.000095

Epoch: 13, Position Loss: 0.007998, Velocity Loss: 0.000168, Position Loss: 0.009014, Velocity Loss: 0.000085

Epoch: 14, Position Loss: 0.007961, Velocity Loss: 0.000166, Position Loss: 0.008901, Velocity Loss: 0.000086

Epoch: 15, Position Loss: 0.007934, Velocity Loss: 0.000165, Position Loss: 0.008995, Velocity Loss: 0.000113

Epoch: 16, Position Loss: 0.007911, Velocity Loss: 0.000164, Position Loss: 0.008803, Velocity Loss: 0.000077

Epoch: 17, Position Loss: 0.007890, Velocity Loss: 0.000163, Position Loss: 0.007344, Velocity Loss: 0.000100

New best val loss: 0.007443627750035375
Epoch: 18, Position Loss: 0.007871, Velocity Loss: 0.000162, Position Loss: 0.009116, Velocity Loss: 0.000108

Epoch: 19, Position Loss: 0.007851, Velocity Loss: 0.000160, Position Loss: 0.008178, Velocity Loss: 0.000085

Epoch: 20, Position Loss: 0.007832, Velocity Loss: 0.000159, Position Loss: 0.007660, Velocity Loss: 0.000091

Epoch: 21, Position Loss: 0.007819, Velocity Loss: 0.000158, Position Loss: 0.007488, Velocity Loss: 0.000102

Epoch: 22, Position Loss: 0.007806, Velocity Loss: 0.000158, Position Loss: 0.008556, Velocity Loss: 0.000106

Epoch: 23, Position Loss: 0.007793, Velocity Loss: 0.000157, Position Loss: 0.009153, Velocity Loss: 0.000165

Epoch: 24, Position Loss: 0.007781, Velocity Loss: 0.000156, Position Loss: 0.008042, Velocity Loss: 0.000061

Epoch: 25, Position Loss: 0.007772, Velocity Loss: 0.000156, Position Loss: 0.008637, Velocity Loss: 0.000103

Epoch: 26, Position Loss: 0.007759, Velocity Loss: 0.000156, Position Loss: 0.007659, Velocity Loss: 0.000083

Epoch: 27, Position Loss: 0.007756, Velocity Loss: 0.000156, Position Loss: 0.007707, Velocity Loss: 0.000101

Epoch: 28, Position Loss: 0.007742, Velocity Loss: 0.000155, Position Loss: 0.007916, Velocity Loss: 0.000070

Epoch: 29, Position Loss: 0.007732, Velocity Loss: 0.000155, Position Loss: 0.007908, Velocity Loss: 0.000140

Epoch: 30, Position Loss: 0.007724, Velocity Loss: 0.000155, Position Loss: 0.008988, Velocity Loss: 0.000110

Epoch: 31, Position Loss: 0.007716, Velocity Loss: 0.000155, Position Loss: 0.008243, Velocity Loss: 0.000100

Epoch: 32, Position Loss: 0.007706, Velocity Loss: 0.000154, Position Loss: 0.007915, Velocity Loss: 0.000156

Epoch: 33, Position Loss: 0.007702, Velocity Loss: 0.000154, Position Loss: 0.009441, Velocity Loss: 0.000111

Epoch: 34, Position Loss: 0.007692, Velocity Loss: 0.000154, Position Loss: 0.007776, Velocity Loss: 0.000152

Epoch: 35, Position Loss: 0.007685, Velocity Loss: 0.000154, Position Loss: 0.007995, Velocity Loss: 0.000135

Epoch: 36, Position Loss: 0.007680, Velocity Loss: 0.000153, Position Loss: 0.008201, Velocity Loss: 0.000087

Epoch: 37, Position Loss: 0.007668, Velocity Loss: 0.000153, Position Loss: 0.008557, Velocity Loss: 0.000110

Epoch: 38, Position Loss: 0.007658, Velocity Loss: 0.000152, Position Loss: 0.007319, Velocity Loss: 0.000181

Epoch: 39, Position Loss: 0.007652, Velocity Loss: 0.000151, Position Loss: 0.007610, Velocity Loss: 0.000119

Epoch: 40, Position Loss: 0.007646, Velocity Loss: 0.000151, Position Loss: 0.008140, Velocity Loss: 0.000183

Epoch: 41, Position Loss: 0.007638, Velocity Loss: 0.000150, Position Loss: 0.006846, Velocity Loss: 0.000179

New best val loss: 0.00702526442182716
Epoch: 42, Position Loss: 0.007637, Velocity Loss: 0.000150, Position Loss: 0.007212, Velocity Loss: 0.000195

Epoch: 43, Position Loss: 0.007630, Velocity Loss: 0.000150, Position Loss: 0.006768, Velocity Loss: 0.000079

New best val loss: 0.006847250508144498
Epoch: 44, Position Loss: 0.007627, Velocity Loss: 0.000150, Position Loss: 0.007409, Velocity Loss: 0.000085

Epoch: 45, Position Loss: 0.007622, Velocity Loss: 0.000149, Position Loss: 0.007089, Velocity Loss: 0.000102

Epoch: 46, Position Loss: 0.007616, Velocity Loss: 0.000149, Position Loss: 0.007061, Velocity Loss: 0.000095

Epoch: 47, Position Loss: 0.007610, Velocity Loss: 0.000149, Position Loss: 0.006463, Velocity Loss: 0.000094

New best val loss: 0.006556672611623071
Epoch: 48, Position Loss: 0.007610, Velocity Loss: 0.000149, Position Loss: 0.006973, Velocity Loss: 0.000096

Epoch: 49, Position Loss: 0.007605, Velocity Loss: 0.000148, Position Loss: 0.007037, Velocity Loss: 0.000099

Epoch: 50, Position Loss: 0.007604, Velocity Loss: 0.000148, Position Loss: 0.006852, Velocity Loss: 0.000088